src/lib.rs

```diff
@@ -14,38 +14,51 @@ fn derive_abomonation(mut s: synstructure::Structure) -> proc_macro2::TokenStrea
     });
 
     let entomb = s.each(|bi| quote! {
-        ::abomonation::Abomonation::entomb(#bi, _write)?;
+        ::abomonation::Entomb::entomb(#bi, _write)?;
     });
 
     let extent = s.each(|bi| quote! {
-        sum += ::abomonation::Abomonation::extent(#bi);
+        sum += ::abomonation::Entomb::extent(#bi);
     });
 
     s.bind_with(|_| synstructure::BindStyle::RefMut);
 
     let exhume = s.each(|bi| quote! {
-        let temp = bytes;
-        bytes = ::abomonation::Abomonation::exhume(#bi, temp)?;
+        bytes = ::abomonation::Exhume::exhume(From::from(#bi), bytes)?;
     });
 
-    s.bound_impl(quote!(abomonation::Abomonation), quote! {
-        #[inline] unsafe fn entomb<W: ::std::io::Write>(&self, _write: &mut W) -> ::std::io::Result<()> {
-            match *self { #entomb }
-            Ok(())
-        }
-        #[allow(unused_mut)]
-        #[inline] fn extent(&self) -> usize {
-            let mut sum = 0;
-            match *self { #extent }
-            sum
+    s.gen_impl(quote! {
+        extern crate abomonation;
+        extern crate std;
+
+        gen unsafe impl abomonation::Entomb for @Self {
+            unsafe fn entomb<W: ::std::io::Write>(&self, _write: &mut W) -> ::std::io::Result<()> {
+                match *self { #entomb }
+                Ok(())
+            }
+
+            #[allow(unused_mut)]
+            fn extent(&self) -> usize {
+                let mut sum = 0;
+                match *self { #extent }
+                sum
+            }
         }
-        #[allow(unused_mut)]
-        #[inline] unsafe fn exhume<'a,'b>(
-            &'a mut self,
-            mut bytes: &'b mut [u8]
-        ) -> Option<&'b mut [u8]> {
-            match *self { #exhume }
-            Some(bytes)
+
+        gen unsafe impl<'de> abomonation::Exhume<'de> for @Self
+            where Self: 'de,
```

Comment:
Synstructure help wanted here: assuming that Self is `Generic<'a, 'b, 'c>`, I would like to enforce that `abomonation::Exhume<'de>` is only implemented for `Generic<'de, 'de, 'de>`.

This `where` clase does not quite do what I want, because it still allows `Self` to be e.g. `&'static Something` when `'de` is shorter-lived than `'static`, which is wrong and can lead to memory unsafety. How do I fix that with synstructure?